server highlights

ZashIn opened the pull request:
Server specific highlights, with the current server as default (global on 0 or DM):
`>[highlight|hl|hladd] <pattern> [server=<this server>]`

Other highlight commands changed accordingly.

Update script for db: [upgrade_db.py](https://github.com/ZashIn/QuoteBot/blob/feat/server-highlight-upgrade/upgrade_db.py)

### TODO
- [x] same pattern overwriting global <-> local.

ZashIn committed:
list all relevant highlights (server id including global)

JonathanFeenstra commented on cogs/highlights.py:
```diff
@@ -37,7 +37,23 @@ def _should_send_highlight(msg: discord.Message, member: discord.Member, query:
     )
 
 
+class OptionalGuildConverter(commands.converter.GuildConverter):
+    async def convert(self, ctx: commands.Context, argument: str) -> discord.Guild | None:
+        if argument == "0":
+            return None
+        return await super().convert(ctx, argument)
+
+
+OptionalCurrentGuild = commands.parameter(
+    default=lambda ctx: ctx.guild,
+    displayed_default="<this server>",
+    converter=OptionalGuildConverter,
+)
```
These should be moved to /core/converters.py. The converter class needs a docstring

ZashIn replied:
Done.

JonathanFeenstra commented on cogs/highlights.py:
```diff
@@ -37,7 +37,23 @@ def _should_send_highlight(msg: discord.Message, member: discord.Member, query:
+
+
 class Highlights(commands.Cog):
+    server_option = re.compile(r" *(?:-s|--server) ?(?P<id>\d+)? *")
```
This regex doesn't seem to be used anymore

JonathanFeenstra commented on cogs/highlights.py:
```diff
@@ -53,11 +69,13 @@ async def _send_highlights(
                 await con.clear_user_highlights(user_id)
             elif not (member := msg.guild.get_member(user_id)):
                 continue
+            elif guild_id and guild_id != msg.guild.id:
```
The first condition (before the `and`) should be made more explicit (`guild_id > 0` or `guild_id is not None`)

ZashIn replied:
`guild_id` can only be an `int`, for which I would usually write `guild_id` instead of `guild_id != 0`. But ok, using `> 0`.

JonathanFeenstra commented on cogs/highlights.py:
```diff
@@ -96,18 +121,21 @@ async def highlight(self, ctx: commands.Context, *, pattern: str) -> None:
-        await ctx.send(f":white_check_mark: **Highlight pattern `{pattern.replace('`', '')}` added.**")
+        await ctx.send(f":white_check_mark: **Highlight pattern `{pattern.replace('`', '')}` added {self._server_text(server)}.**")
+
+    def _server_text(self, server, pre="", no_server="globally"):
```
This method still needs type hints. In the code, it should be called `guild` instead of `server` unless it's visible to the user in which case `server` would be less confusing

ZashIn replied:
Oversight after refactoring  Using the terms `server`and `guild` interchangeable in code is a bit confusing, true.
But I would stick to the parameter name "server" for commands, since guild would confuse most users.
The rest is fixed.

JonathanFeenstra commented on cogs/highlights.py:
```diff
@@ -116,40 +144,53 @@ async def highlightlist(self, ctx: commands.Context) -> None:
+            await ctx.send(f":x: **You don't have any Highlights{self._server_text(server, ' ', '')}.**")
+
+    def _hightlight_table_str(self, highlights: tuple[tuple[str, int], ...], ctx: commands.Context) -> str:
+        hl_table = []
+        max_pattern_len = len("pattern")
+        for pattern, gid in highlights:
+            max_pattern_len = max(max_pattern_len, len(pattern))
+            gid = f"{gid} : {ctx.bot.get_guild(gid).name}" if gid else "0"
+            hl_table.append((pattern.replace('`', ''), gid))
+        header = f"*`{'pattern':<{max_pattern_len}}  server`*\n"  # aligned italic header
+        return header + "\n".join(f"`{pattern:<{max_pattern_len}}  {gid}`" for pattern, gid in hl_table)
```
For better readability, you should write variable names in full, not abbreviated (`highlight_table`, `max_pattern_length`, `guild_id`)

ZashIn replied:
Obsolete: replaced by guild list:

global `(0)`:
> `global-pattern`

server `Bots test (802975314103238656)`:
>  `local-.pattern`
> `local-.pattern2`

JonathanFeenstra commented on core/persistence.py:
```diff
@@ -141,33 +141,58 @@ async def delete_blocked_id(self, blocked_id: int) -> None:
+            rows = await self.execute_fetchall(
+                "SELECT query, guild_id FROM highlight WHERE user_id = ? AND query LIKE ?", (user_id, f"{prefix}%")
+            )
+        return tuple(tuple(str(i) for i in row) for row in rows)
```
It doesn't really make sense to convert the `guild_id` to a string here since it's left as an `int` everywhere else. Rename `i` to `value` for clarity

ZashIn replied:
Agreed, changed the other list commands to return specific (typed) values aswell.

ZashIn committed:
highlightlist: grouped by servers

highlightlist: formatting bold

ZashIn committed:
Refactor: inline / simplify helper functions, renaming

ZashIn committed:
Converters: OptionalGuildConverter, OptionalCurrentGuild = no guild / global with 0 or "guild"

ZashIn committed:
local <-> global highlight warning / overwrite

ZashIn committed:
persistence.py: consistent typing changes for py 3.10 (matching highlights.py)

ZashIn committed:
DB: add guild_id to highlight table

ZashIn committed:
highlight: with server/guild + insert values with guild_id into DB

ZashIn committed:
on_message: + guild_id

ZashIn committed:
highlightlist & fetch_user_highlights: + server/guild

ZashIn committed:
highlightremove: + server/guild

ZashIn committed:
highlightclear: + server/guild

ZashIn commented:
All requested changes incorporated, history separated + cleaned.
Ready to be merged from my side.

JonathanFeenstra commented on core/converters.py:
```diff
@@ -64,3 +64,19 @@ async def convert(self, ctx: commands.Context, argument: str) -> TextChannelOrTh
+    displayed_default="<this server>",
+    converter=OptionalGuildConverter,
+)
+"""`commands.CurrentGuild` with :class:`OptionalGuildConverter`, returning None without a current guild."""
```
Does this docstring work? Don't you need [`typing.Annotated`](https://docs.python.org/3/library/typing.html#typing.Annotated) here?

ZashIn replied:
Docstrings after variables is supported by several tools & styles, including Sphinx, VSCode...
What docstring style do you prefer?
I think `typing.Annotated` has different syntax.

JonathanFeenstra replied:
If VSCode supports it then it's ok

JonathanFeenstra commented on cogs/highlights.py:
```diff
@@ -69,11 +69,21 @@ async def _send_highlights(
                     continue
         await con.commit()
 
+    def _get_guild_id(self, server: discord.Guild | None) -> int:
```
This method doesn't use the `self` argument so it should be a function instead of a method. Actually it doesn't do enough to justify creating a function either, this should really just be in-lined.

ZashIn replied:
Fixed: inlined

JonathanFeenstra commented on cogs/highlights.py:
```diff
@@ -69,11 +69,21 @@ async def _send_highlights(
+    def _get_guild_id(self, server: discord.Guild | None) -> int:
+        return server.id if server else 0
+
+    def _server_text(self, server, pre="", no_server="globally"):
```
This should be a function and the names should be clarified for better readability (`pre` -> `prefix`). `if server else no_server` should not be part of the function but in-lined where necessary. It also needs type hints.

ZashIn replied:
Fixed: simplified (optional args removed), moved to function "_for_guild_str(guild)".

JonathanFeenstra reviewed (changes requested):
I don't like the way the commit history currently looks, these are some guidelines to improve it:

- All changes that you'd expect to see together (like adding the `OptionalCurrentGuild` parameter and actually using it somewhere in the code) should be part of the same commit. 
- Commit messages should briefly describe what the change does in present tense (e.g. "Add option to make highlight commands server-specific").
-  Commits that correct issues in previous commits should be rebased and fixed in the original commit (I always use the fixup-option in [interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)). This makes it easier to review code commit-by-commit, otherwise I might comment on something that has already been fixed in a later commit.

ZashIn commented:
> I don't like the way the commit history currently looks, these are some guidelines to improve it

Thanks for the tips and sorry for the messy commits, I am not familiar with discord bot / api, so had to play around a bit.
It was also difficult to separate the commits properly, since most of the DB / `persistent.py` changes affect all commands in the end.

It would be easiest to just do a squash merge with "Add option to make highlight commands server-specific".
Otherwise I would have to rewrite the entire history again to satisfy all history requirements, besides simple conflict free squashes and rephrasing messages.
Even then the bot would probably still not run without error in the intermediate commits (no proper test driven development here ofc)...

I rather invest the time in something else, e.g. channel filtering and make propper commits in the next PR.

The only point that is not clear to me is the following:
> * Commits that correct issues in previous commits should be rebased and fixed in the original commit (I always use the fixup-option in [interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)). This makes it easier to review code commit-by-commit, otherwise I might comment on something that has already been fixed in a later commit.

Do you mean a rebase for changes before or after push? I.e. 1 commit for all fixes (squashed / fixed up together) or a rebase edit of original commits & force push?
Local rebasing before pushing goes without saying.

To elaborate:
There are two conflicting lines of though here:
1. Never change (force push / rebase) published commits to avoid pull conflicts for others.
    In other words: always add fixes as new commits after push / publishing, only rewrite history (rebase squash / fixup) for local changes.
2. Keep history clean, esp. in master (either with merge commits or rebase)
And for the merge into master accordingly: either squash, "Rebase and merge" (fast forward `-ff`) or create merge commit (`--no-ff`)

Originally, I followed 1. and added the changes as separate commits (~1 per fix), in case you checked the PR out.
But after your feedback on Discord ("separate the commits into related changes"), I rewrote the full PR history into related changes, e.g. ~ one commit per sub-command change.

What do you prefer for PRs? One or more commits for fixes (and e.g. review them as range of commits) or a rebase and force push?
Imo the latter makes it harder to follow the changes before/after review (e.g. last commit / push removing the commit referenced in review).
In the former case an option to make the re-review easier would be to do it over a range of commits maybe?

And just to be on the same page for the next PR:
> * All changes that you'd expect to see together (like adding the OptionalCurrentGuild parameter and actually using it somewhere in the code) should be part of the same commit

That would mean squashing e.g.:
- first 3 commits: Converters, DB change and first highlight commit
- `highlightlist: grouped by servers`: squash into first highlight commit
- `code: naming server vs guild`

JonathanFeenstra commented:
> It would be easiest to just do a squash merge with "Add option to make highlight commands server-specific".

That's fine, if rebasing is going to be too much of a hassle I wouldn't bother spending too much time on it.

What I meant was option 2, clean the commit history and force push (using `--force-with-lease`). Not changing published commits is indeed good practice for long-living branches like master and develop (for repos that use it), but less important for feature/fix branches in my opinion. It's more important to keep the commit history organised for easy reviewing and to find changes more easily after merging.

Keeping track of what was changed before and after the review is not a problem, GitHub adds "Compare" buttons in the PR every time you push.

JonathanFeenstra commented on cogs/highlights.py:
```diff
@@ -95,19 +105,38 @@ async def highlight(self, ctx: commands.Context, *, pattern: str) -> None:
+                    global_overwritten = True
+            elif guilds := await con.fetch_user_highlight_guilds(user_id, pattern, exclude_global_guild=True):
+                # Warning about server highlights
+                guilds_str = "\n".join([f"`{gid} : {ctx.bot.get_guild(server.id if server else 0).name}`" for gid in guilds])
```
```suggestion
                guilds_str = "\n".join([f"`{guild_id} : {ctx.bot.get_guild(server.id if server else 0).name}`" for guild_id in guilds])
```

JonathanFeenstra commented on cogs/highlights.py:
```diff
@@ -95,19 +105,38 @@ async def highlight(self, ctx: commands.Context, *, pattern: str) -> None:
                 await ctx.send(":x: **Highlight limit exceeded.**")
+            guild_id = server.id if server else 0
+            global_overwritten = False
+            if guild_id:
```
```suggestion
            if guild_id != 0:
```

JonathanFeenstra commented on cogs/highlights.py:
```diff
@@ -53,11 +57,13 @@ async def _send_highlights(
                 await con.clear_user_highlights(user_id)
             elif not (member := msg.guild.get_member(user_id)):
                 continue
+            elif guild_id > 0 and guild_id != msg.guild.id:
```
```suggestion
            elif guild_id != 0 and guild_id != msg.guild.id:
```

JonathanFeenstra commented on core/persistence.py:
```diff
@@ -141,47 +142,93 @@ async def delete_blocked_id(self, blocked_id: int) -> None:
+    async def fetch_user_highlights_starting_with(
+        self, user_id: int, prefix: str, guild_id: int = 0
+    ) -> tuple[tuple[str, int], ...]:
+        if guild_id:
```
```suggestion
        if guild_id != 0:
```

JonathanFeenstra commented on core/persistence.py:
```diff
@@ -141,47 +142,93 @@ async def delete_blocked_id(self, blocked_id: int) -> None:
-    async def clear_user_highlights(self, user_id: int) -> None:
-        await self.execute("DELETE FROM highlight WHERE user_id = ?", (user_id,))
+    async def clear_user_highlights(self, user_id: int, guild_id: int = 0) -> None:
+        if guild_id:
```
```suggestion
        if guild_id != 0:
```

ZashIn replied:
Ok, good, I see you prefer this explicitly in general, will keep it in mind.

Just a thought:
With restricted types / type checking it makes no difference, besides arguably readability. (Might type checking via mypy e.g. then be a good option in the workflow to enforce it?)
Without strict typing (since it is technically optional in python), just `if guild_id` would catch e.g. passed `None` values, instead of passing `NULL` to the query, resulting in deleting nothing (or returning nothing above).
But I understand your point.

JonathanFeenstra replied:
I use Pyright for type checking, but type safety is not enforced in the CI since there are some places where types aren't correctly recognised (for example when converters are used as type hints instead of the actual type). The explicit comparisons are just a matter of taste. In my opinion, it's more readable and easier to understand, especially if you're coming from a different programming language.

JonathanFeenstra reviewed (changes requested):
Since it's getting squashed anyway, it's ok to commit these suggestions directly in the PR.

JonathanFeenstra committed:
Update cogs/highlights.py

JonathanFeenstra committed:
Update cogs/highlights.py

JonathanFeenstra committed:
Update cogs/highlights.py

JonathanFeenstra committed:
Update core/persistence.py

JonathanFeenstra committed:
Update core/persistence.py

JonathanFeenstra committed:
Format with black
